pkg/download/consistent_hashing.go

```diff
@@ -0,0 +1,263 @@
+	totalSlices := fileSize / m.SliceSize
+	if fileSize%m.SliceSize != 0 {
+		totalSlices++
+	}
+
+	errGroup, ctx := errgroup.WithContext(ctx)
+	errGroup.Go(func() error {
+		return m.downloadChunk(firstChunkResp, data[0:m.minChunkSize()])
+	})
+
+	// we subtract one because we've already got firstChunkResp in flight
+	concurrency := m.maxConcurrency() - 1
+	if concurrency <= int(totalSlices) {
+		// special case: we should just download each slice in full and rely on the semaphore to manage concurrency
+		concurrency = int(totalSlices)
+	}
+
+	chunksPerSlice := EqualSplit(int64(concurrency), totalSlices)
+	if m.minChunkSize() == m.SliceSize {
+		// firstChunkResp will download the whole first chunk in full;
+		// we set slice 0 to a special value of 0 so we skip it later
+		chunksPerSlice = append([]int64{0}, EqualSplit(int64(concurrency), totalSlices-1)...)
+	}
+
+	logger.Debug().Str("url", urlString).
+		Int64("size", fileSize).
+		Int("concurrency", m.maxConcurrency()).
+		Ints64("chunks_per_slice", chunksPerSlice).
+		Msg("Downloading")
+
+	for slice, numChunks := range chunksPerSlice {
+		if numChunks == 0 {
+			// this happens if we've already downloaded the whole first slice
+			continue
+		}
+		start := m.SliceSize * int64(slice)
+		sliceSize := m.SliceSize
+		if slice == 0 {
+			start = firstChunkResp.ContentLength
+			sliceSize = sliceSize - firstChunkResp.ContentLength
+		}
+		if slice == int(totalSlices)-1 {
+			sliceSize = (fileSize-1)%m.SliceSize + 1
+		}
+		if sliceSize/numChunks < m.minChunkSize() {
+			// reset numChunks to respect minChunkSize
+			numChunks = sliceSize / m.minChunkSize()
+			// although we must always have at least one chunk
+			if numChunks == 0 {
+				numChunks = 1
+			}
+		}
+		chunkSizes := EqualSplit(sliceSize, numChunks)
+		for _, chunkSize := range chunkSizes {
+			end := start + chunkSize - 1
+
+			logger.Debug().Int64("start", start).Int64("end", end).Msg("starting request")
+			resp, err := m.doRequest(ctx, start, end, urlString)
+			if err != nil {
+				return nil, -1, err
+			}
+
+			dataSlice := data[start : end+1]
+			errGroup.Go(func() error {
+				return m.downloadChunk(resp, dataSlice)
+			})
+
+			start = start + chunkSize
+		}
+	}
+
+	if err := errGroup.Wait(); err != nil {
+		return nil, -1, err // return the first error we encounter
+	}
+
+	buffer := bytes.NewBuffer(data)
+	return buffer, fileSize, nil
+}
+
+func (m *ConsistentHashingMode) doRequest(ctx context.Context, start, end int64, urlString string) (*http.Response, error) {
+	logger := logging.GetLogger()
+	if m.Semaphore != nil {
+		err := m.Semaphore.Acquire(ctx, 1)
+		if err != nil {
+			return nil, err
+		}
+	}
+	req, err := http.NewRequestWithContext(ctx, "GET", urlString, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to download %s: %w", req.URL.String(), err)
+	}
+	err = m.consistentHashIfNeeded(req, start, end)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", start, end))
```

Comment:
NIT: I'd love to get "Range" into a constant just so future things can't toyp it, same with bytes=.

Not worth changing here.